File: avancado-intermediario/Python/Q2/pessoa.py
class Pessoa:
    def __init__(self, nome: str, pai: 'Pessoa' = None, mae: 'Pessoa' = None):
        self.__nome = nome
        self.__pai = pai
        self.__mae = mae

    def getNome(self) -> str:
        return self.__nome

    def getPai(self) -> 'Pessoa':
        return self.__pai

    def getMae(self) -> 'Pessoa':
        return self.__mae

    def verificarIgualdadeSemantica(self, outraPessoa: 'Pessoa') -> bool:
        if outraPessoa is None:
            return False
        mesmoNome = self.__nome == outraPessoa.getNome()
        mesmoPai = self.__pai is outraPessoa.getPai()
        mesmaMae = self.__mae is outraPessoa.getMae()
        return mesmoNome and mesmoPai and mesmaMae

    def verificarIrmandade(self, outraPessoa: 'Pessoa') -> bool:
        if outraPessoa is None or self.verificarIgualdadeSemantica(outraPessoa):
            return False
        
        mesmoPai = self.__pai is not None and self.__pai.verificarIgualdadeSemantica(outraPessoa.getPai())
        mesmaMae = self.__mae is not None and self.__mae.verificarIgualdadeSemantica(outraPessoa.getMae())
        
        return mesmoPai or mesmaMae

    def verificarAncestralidade(self, possivelAncestral: 'Pessoa') -> bool:
        if possivelAncestral is None:
            return False

        if (self.__pai is not None and self.__pai.verificarIgualdadeSemantica(possivelAncestral)) or \
           (self.__mae is not None and self.__mae.verificarIgualdadeSemantica(possivelAncestral)):
            return True

        ancestralPeloPai = self.__pai is not None and self.__pai.verificarAncestralidade(possivelAncestral)
        ancestralPelaMae = self.__mae is not None and self.__mae.verificarAncestralidade(possivelAncestral)

        return ancestralPeloPai or ancestralPelaMae

    def __str__(self) -> str:
        nomePai = self.__pai.getNome() if self.__pai is not None else "desconhecido(a)"
        nomeMae = self.__mae.getNome() if self.__mae is not None else "desconhecido(a)"
        return "Nome: " + self.__nome + ", Pai: " + nomePai + ", Mãe: " + nomeMae

File: avancado-intermediario/Java/Q2/Pessoa.java
public class Pessoa {
    private String nome;
    private Pessoa pai;
    private Pessoa mae;

    public Pessoa(String nome, Pessoa pai, Pessoa mae) {
        this.nome = nome;
        this.pai = pai;
        this.mae = mae;
    }

    public Pessoa(String nome) {
        this.nome = nome;
        this.pai = null;
        this.mae = null;
    }

    public String getNome() {
        return nome;
    }

    public boolean verificarIgualdadeSemantica(Pessoa outraPessoa) {
        if (outraPessoa == null) {
            return false;
        }

        boolean mesmoNome = this.nome.equals(outraPessoa.getNome());
        boolean mesmoPai = (this.pai == outraPessoa.pai);
        boolean mesmaMae = (this.mae == outraPessoa.mae);

        return mesmoNome && mesmoPai && mesmaMae;
    }

    public boolean verificarIrmandade(Pessoa outraPessoa) {
        if (outraPessoa == null || this.verificarIgualdadeSemantica(outraPessoa)) {
            return false; 
        }

        boolean mesmoPai = (this.pai != null && this.pai.verificarIgualdadeSemantica(outraPessoa.pai));
        boolean mesmaMae = (this.mae != null && this.mae.verificarIgualdadeSemantica(outraPessoa.mae));

        return mesmoPai || mesmaMae;
    }

    public boolean verificarAncestralidade(Pessoa possivelAncestral) {
        if (possivelAncestral == null) {
            return false;
        }

        if ((this.pai != null && this.pai.verificarIgualdadeSemantica(possivelAncestral)) ||
                (this.mae != null && this.mae.verificarIgualdadeSemantica(possivelAncestral))) {
            return true;
        }

        boolean ancestralPeloPai = (this.pai != null && this.pai.verificarAncestralidade(possivelAncestral));
        boolean ancestralPelaMae = (this.mae != null && this.mae.verificarAncestralidade(possivelAncestral));

        return ancestralPeloPai || ancestralPelaMae;
    }

    @Override
    public String toString() {
        String nomePai = (pai != null) ? pai.getNome() : "desconhecido(a)";
        String nomeMae = (mae != null) ? mae.getNome() : "desconhecido(a)";
        return "Nome: " + nome + ", Pai: " + nomePai + ", Mãe: " + nomeMae;
    }
}
